p2p/nodestate/nodestate.go

```diff
@@ -646,53 +659,104 @@ func (ns *NodeStateMachine) SetState(n *enode.Node, setFlags, resetFlags Flags,
+							cb(n, Flags{setup: ns.setup}, v, nil)
+						}
 					}
 				}
 			}
 		}
 	}
+	callNow := node.pendingCallbacks == nil
+	node.pendingCallbacks = append(node.pendingCallbacks, callback)
+	ns.callbackCount++
+	list := node.pendingCallbacks
+	ns.lock.Unlock()
+	// call state update subscription callbacks without holding the mutex
+	if callNow {
+		ns.processCallbacks(node, list)
+	}
+}
+
+// processCallbacks runs pending callbacks of a given node in a guaranteed correct order.
+// Callbacks resulting from a state/field change performed in a previous callback are always
+// put at the end of the pending list and therefore processed after all callbacks resulting
+// from the previous state/field change.
+func (ns *NodeStateMachine) processCallbacks(node *nodeInfo, list []func()) {
+	for list != nil {
+		for _, cb := range list {
+			cb()
+		}
+		ns.lock.Lock()
+		node.pendingCallbacks = node.pendingCallbacks[len(list):]
+		if len(node.pendingCallbacks) == 0 {
+			node.pendingCallbacks = nil
+		}
+		ns.callbackCount -= len(list)
+		if ns.callbackCount == 0 {
+			ns.callbackWait.Signal()
+		}
+		list = node.pendingCallbacks
+		ns.lock.Unlock()
+	}
 }
 
 // offlineCallbacks calls state update callbacks at startup or shutdown
 func (ns *NodeStateMachine) offlineCallbacks(start bool) {
 	for _, cb := range ns.offlineCallbackList {
-		for _, sub := range ns.stateSubs {
-			offState := offlineState & sub.mask
-			onState := cb.state & sub.mask
-			if offState != onState {
-				if start {
-					sub.callback(cb.node, Flags{mask: offState, setup: ns.setup}, Flags{mask: onState, setup: ns.setup})
-				} else {
-					sub.callback(cb.node, Flags{mask: onState, setup: ns.setup}, Flags{mask: offState, setup: ns.setup})
-				}
-			}
-		}
-		for i, f := range cb.fields {
-			if f != nil && ns.fields[i].subs != nil {
-				for _, fsub := range ns.fields[i].subs {
+		cb := cb
+		callback := func() {
+			for _, sub := range ns.stateSubs {
+				offState := offlineState & sub.mask
+				onState := cb.state & sub.mask
+				if offState != onState {
 					if start {
-						fsub(cb.node, Flags{mask: offlineState, setup: ns.setup}, nil, f)
+						sub.callback(cb.node.node, Flags{mask: offState, setup: ns.setup}, Flags{mask: onState, setup: ns.setup})
 					} else {
-						fsub(cb.node, Flags{mask: offlineState, setup: ns.setup}, f, nil)
+						sub.callback(cb.node.node, Flags{mask: onState, setup: ns.setup}, Flags{mask: offState, setup: ns.setup})
+					}
+				}
+			}
+			for i, f := range cb.fields {
+				if f != nil && ns.fields[i].subs != nil {
+					for _, fsub := range ns.fields[i].subs {
+						if start {
+							fsub(cb.node.node, Flags{mask: offlineState, setup: ns.setup}, nil, f)
+						} else {
+							fsub(cb.node.node, Flags{mask: offlineState, setup: ns.setup}, f, nil)
+						}
 					}
 				}
 			}
 		}
+		ns.lock.Lock()
+		if cb.node.pendingCallbacks != nil {
+			panic("Fatal: unfinished callback")
+		}
+		cb.node.pendingCallbacks = []func(){callback}
```

Comment:
Why we always assign the `callback` to `pendingCallbacks`?

Reply:
Right, the first one is unnecessary to put there.